alibi/explainers/similarity/backends/tensorflow/base.py

```diff
@@ -79,3 +98,11 @@ def argmax(X: tf.Tensor, dim=-1) -> tf.Tensor:
         """Returns the index of the maximum value in a tensor."""
         X = tf.math.argmax(X, axis=dim)
         return X
+
+    @staticmethod
+    def get_non_trainable(model: keras.Model) -> List[Union[int, str]]:
+        """Checks if all layers in a model are trainable.
+
+        Note: batch normalization layers are ignored as they are not trainable by default.
+        """
+        return [getattr(layer, 'name', i) for i, layer in enumerate(model.layers) if not layer.trainable]
```

Comment:
This approach does not work as for the `pytorch` backend.
For example, consider the model:

```python
model = keras.Sequential([
    keras.layers.Dense(10),
    keras.layers.Dense(20),
    keras.Sequential([
        keras.layers.Dense(30),
        keras.layers.Dense(40),
    ])    
])

model.build((None, 10))
```

Calling `model.layers` will return 

```
[<keras.layers.core.dense.Dense at 0x7f54d4332430>,
 <keras.layers.core.dense.Dense at 0x7f54d43323d0>,
 <keras.engine.sequential.Sequential at 0x7f54d44dea30>]
```

Setting on layer inside the sequential layer to non-trainable:
``` python
model.layers[-1].layers[0].trainable = False
```

If we call your function, it returns an empty list, which is incorrect.
``` python
get_non_trainable(model)
```

If we call it with the sequential layer, then it returns the correct answer.
```python
get_non_trainable(model.layers[-1])
```

This shows that the function does not work for nested layers, which are very common. Not sure if I am missing something and if the problem is addressed somewhere else?

Reply:
Yeah, good catch! :man_facepalming: It's better if I just iterate through the `model.non_trainable_weights` here!

Reply:
It's a similar issue for `tensorflow` models which have nesting, e.g. any transformer model (see the model for our text app as an example). I think we should try a recursive approach.

Reply:
I think @RobertSamoilescu misspoke no? The above is a `Keras` model?